Configure OQT using files or environment variables

matthiasschaub opened the pull request:
### Description
Configure OQT using files or environment variables.

All configuration can be done using environment variables.

All configuration can also be done by providing a configuration file (`yaml` format). 
The default configuration file path is `workers/config/config.yaml`. The path to the configuration can be changed by setting the path in the environment variable `OQT_CONFIG`. 
A sample configuration file is provided at `workers/config/config.yaml`.

### Corresponding issue
- None

### Checklist
- [x] I have updated my branch to `main` (e.g. through `git rebase main`)
- [x] My code follows the [style guide](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#style-guide) and was checked with [pre-commit](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#pre-commit) before committing
- [x] I have commented my code
- [x] I have added sufficient unit and integration [tests](https://github.com/GIScience/ohsome-quality-analyst/blob/main/docs/development_setup.md#tests)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CHANGELOG.md)
- [x] Update import statements of definitions.py
- [x] Write configuration docs
    - [x] Provide a table with all possible configuration variables
    - [x] Provide prose on how to setup config files
- [x] Clean up commit history
- [x] Add configuration for semaphore (#307)

Open question:
- Should configuration variables be made available as globals through definitions.py? -> No. Two functions to get either all config vars or a single value for a given key are provided.
- Tests for config are very minmal. Should more time be spent to implement further tests?

joker234 commented on workers/tests/unittests/test_cli.py:
```diff
@@ -57,6 +57,9 @@ def test_create_all_indicators_invlid_opts(self, mo):
     @mock.patch("ohsome_quality_analyst.oqt.create_indicator", mock.AsyncMock())
+    @mock.patch(
+        "ohsome_quality_analyst.geodatabase.client.get_feature_ids", mock.AsyncMock()
+    )
```
What was changed to make this mock patch necessary?

matthiasschaub replied:
Good catch. Nothing changed to make it necessary. I removed this mock.

joker234 commented on workers/tests/unittests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+
+
+class TestConfig(unittest.TestCase):
+    @mock.patch.dict("os.environ", {}, clear=True)
+    def test_get_config_path_empty_env(self):
+        self.assertEqual(
+            config.get_config_path(),
+            os.path.join(
+                os.path.dirname(
+                    inspect.getfile(config),
+                ),
+                "config.yaml",
+            ),
+        )
```
I would write this test differently and not copy the definition of the method into the unittest. I would compare it explicitly with the path string you are using.

matthiasschaub replied:
The expected path is now build relative to the test file.

joker234 commented on workers/tests/unittests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+    def test_get_config_path_set_env(self):
+        self.assertEqual(
+            config.get_config_path(),
+            os.path.join(tempfile.gettempdir(), "config.yaml"),
```
see above. It shouldn't be *exactly* the same implementation. I would use strings here.

joker234 replied:
Okay maybe I just don't get why you're using the temp directory and not just any string/path?

joker234 commented on workers/tests/unittests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+        self.assertIsInstance(cfg, dict)
+        self.assertTrue(cfg)
+
+    def test_check_missing_conf_keys_compete(self):
```
```suggestion
    def test_check_missing_conf_keys_complete(self):
```

matthiasschaub replied:
This test has been removed since the function to be tested has been removed.

joker234 commented on workers/ohsome_quality_analyst/config/config.py:
```diff
@@ -0,0 +1,168 @@
+    return cfg
+
+
+def check_missing_conf_keys(cfg) -> dict:
+    """Check for missing configuration keys.
+
+    Returns:
+        dict: Input dict without changes
+    Raises:
+        KeyError: In case configuration key is missing from configuration dictionary.
+    """
```
What is the semantic difference between a missing key and a key explicitly set to `None`/`none`? I would say there isn't any and therefore `None`-keys should be ignored.

matthiasschaub replied:
This validation function has been removed in favor of default values if config is not provided.

joker234 commented on workers/tests/unittests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+        self.assertTrue(cfg)
+
+    def test_check_missing_conf_keys_compete(self):
+        cfg = {
+            "postgres_host": None,
+            "postgres_port": None,
+            "postgres_db": None,
+            "postgres_user": None,
+            "postgres_password": None,
+            "data_dir": None,
+            "geom_size_limit": None,
+            "log_level": None,
+            "ohsome_api": None,
+            "test_db": None,
+            "user_agent": None,
+            "datasets": None,
+        }
```
I think `None` should be replaced through some values. See comment to `config.check_missing_conf_keys(…)`.

matthiasschaub replied:
Configuration setup has been changed to always use default values if no configuration has been provided

joker234 commented on workers/tests/unittests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+        with self.assertRaises(KeyError):
+            config.check_missing_conf_keys(cfg)
+
+    def test_get_data_dir_unset_env(self):
+        data_dir = config.get_default_data_dir()
+        expected = os.path.join(
+            os.path.dirname(
+                os.path.abspath(__file__),
+            ),
+            "..",
+            "..",
+            "..",
+            "data",
+        )
+        self.assertTrue(os.path.samefile(data_dir, expected))
```
This tests the same implementation in the test and the method. I would suggest to use a string instead.

matthiasschaub replied:
But then it is only valid for linux systems is it not.

joker234 commented on workers/tests/unittests/test_definitions.py:
```diff
@@ -73,41 +54,6 @@ def test_get_raster_dataset_undefined(self):
-        )
-        self.assertTrue(os.path.samefile(data_dir, expected))
-
-    def test_get_data_dir_set_env(self):
-        with tempfile.TemporaryDirectory() as tmpdirname:
-            custom_data_dir = os.path.join(tmpdirname, "oqt-custom")
-            os.mkdir(custom_data_dir)
-            with mock.patch.dict(
-                "os.environ",
-                {"OQT_DATA_DIR": custom_data_dir},
-                clear=True,
-            ):
-                data_dir = definitions.get_data_dir()
-            self.assertEqual(data_dir, custom_data_dir)
```
I couldn't really find a replacement for this test. Is this not necessary anymore?

joker234 commented on workers/tests/unittests/test_raster.py:
```diff
@@ -20,23 +22,20 @@ def setUp(self):
-    )
-    @mock.patch("os.path.exists", return_value=True)
+    @mock.patch("ohsome_quality_analyst.raster.client.DATA_DIR", tempfile.gettempdir())
     def test_get_raster_path(self, *args):
+        Path(
+            tempfile.gettempdir() + "/GHS_BUILT_LDS2014_GLOBE_R2018A_54009_1K_V2_0.tif"
+        ).touch()
+
         self.assertEqual(
             raster_client.get_raster_path(self.raster_dataset),
             os.path.join(
-                "test_dir", "GHS_BUILT_LDS2014_GLOBE_R2018A_54009_1K_V2_0.tif"
+                tempfile.gettempdir(),
+                "GHS_BUILT_LDS2014_GLOBE_R2018A_54009_1K_V2_0.tif",
             ),
         )
```
Tests shouldn't mess with files in my `/tmp`. Please always create a tempdir/-file and remove it after the tests are done.

matthiasschaub replied:
Now a tmp dir get created for this test

joker234 reviewed (changes requested):
Thanks for the PR. I have made some comments/questions, mostly about the details of the tests and the config design.

Two general remarks:
1. Is there any config variable without a default and mandatory to be set by the user? If no, I would really prefer to not having the config file mandatory.
2. The tests cannot depend on the local configuration. That doesn't make the tests independent, which they should be.

matthiasschaub commented on workers/ohsome_quality_analyst/config/sample.config.yaml:
```diff
@@ -0,0 +1,20 @@
+---
```
Even if documentation about those exists in the docs directory? There are nor mandatory options.

matthiasschaub commented on workers/ohsome_quality_analyst/config/sample.config.yaml:
```diff
@@ -0,0 +1,20 @@
+ohsome_api: https://api.ohsome.org/v1/
+test_db: False
+# Default is 'ohsome-quality-analyst/{version}'
+user_agent: null
```
No more `null` anymore!

matthiasschaub commented on workers/Dockerfile:
```diff
@@ -25,4 +25,5 @@ RUN python -m poetry install --no-ansi --no-interaction --no-root
 
 # copy all the other files and install the project
 COPY --chown=oqt:oqt . .
+RUN cp ohsome_quality_analyst/config/sample.config.yaml ohsome_quality_analyst/config/config.yaml
```
It is not anymore. I added a function to get default values for every config key. I reverted this change.

matthiasschaub commented on workers/ohsome_quality_analyst/config/config.py:
```diff
@@ -0,0 +1,168 @@
+    """
+    with open(path, "r") as f:
+        cfg = yaml.safe_load(f)
+    return {k: v for k, v in cfg.items() if v is not None}
```
Removed the sentence in docstring hinting that this is the case.

matthiasschaub commented on workers/ohsome_quality_analyst/config/sample.config.yaml:
```diff
@@ -0,0 +1,20 @@
+postgres_user: oqt
+postgres_password: oqt
+# Default is the root of the repo
+data_dir: null
```
No more `null` values present anymore

matthiasschaub commented on workers/ohsome_quality_analyst/definitions.py:
```diff
@@ -50,6 +28,15 @@ class RasterDataset:
 
 
+CONFIG = get_config()
+# Dataset names and fid fields which are available in the Geodatabase
+DATASETS = MappingProxyType(CONFIG["datasets"])
+DATA_DIR = CONFIG["data_dir"]
+GEOM_SIZE_LIMIT = CONFIG["geom_size_limit"]
+OHSOME_API = CONFIG["ohsome_api"].rstrip("/")
+USER_AGENT = CONFIG["user_agent"]
```
There are not really necessary. I removed them and replaced their usages by accessing keys in `CONFIG` directly.

matthiasschaub commented:
> Thanks for the PR. I have made some comments/questions, mostly about the details of the tests and the config design.
> 
> Two general remarks:
> 
>     1. Is there any config variable without a default and mandatory to be set by the user? If no, I would really prefer to not having the config file mandatory.
> 
>     2. The tests cannot depend on the local configuration. That doesn't make the tests independent, which they should be.

Thanks for the throughout review so far!
After I addressed your review comments following statements about your general points are true:
1. There is now mandatory config variable to be set and a config file is not mandatory
2. The tests are not depending on a present config file anymore.

joker234 commented on docs/development_setup.md:
```diff
@@ -80,30 +80,11 @@ pre-commit install  # Install pre-commit hooks.
 
-
-#### Local database
+For all possible configuration parameter please refer to the [configuration documentation](/docs/configuration.md).
```
Why not? The configuration shouldn't have anything to do with the image but only being necessary when starting a container.

joker234 commented on workers/tests/unittests/test_config.py:
```diff
@@ -31,15 +51,34 @@ def test_get_config_path_set_env(self):
+    def test_config_default(self):
+        cfg = config.load_config_default()
+        self.assertIsInstance(cfg, dict)
+        self.assertListEqual(self.keys, list(cfg.keys()))
```
This comment applies to future occurrences as well:
Is the order of the keys relevant? `assertListEqual` checks the lists, which are ordered.

matthiasschaub replied:
Order is not necessary. Replaced with just `assertEqual`.

joker234 replied:
Which doesn't change my question, since:

> These methods are used by default when comparing lists or tuples with [assertEqual()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual).

https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertListEqual

matthiasschaub replied:
Now both lists to be compared are sorted beforehand.

joker234 reviewed (commented):
**Changelog missing!**

matthiasschaub commented on workers/ohsome_quality_analyst/config/config.py:
```diff
@@ -49,62 +73,30 @@ def load_config_from_env() -> dict:
-    return cfg
+def get_config_value(key: str) -> Union[str, int, dict]:
+    config = get_config()
+    return config[key]
```
No not really. This error should only occur during development and that not often and easily fixable.

matthiasschaub commented on workers/ohsome_quality_analyst/definitions.py:
```diff
@@ -123,15 +100,6 @@ class RasterDataset:
     ("TagsRatio", "jrc_bridge_count"),
     ("TagsRatio", "jrc_mass_gathering_sites_count"),
 )
-OHSOME_API = os.getenv("OHSOME_API", default="https://api.ohsome.org/v1/").rstrip("/")
```
Fixed. This is now reflected in both docker-compose files

matthiasschaub commented on docs/configuration.md:
```diff
@@ -1,27 +1,32 @@
+| Datasets and Features IDs   | -                         | `datasets`              | `[{"regions": {"default": "ogc_fid"}}]`             | Dataset and Features Ids available in the database (see description below)   |
+| Geometry Size Limit (km²)   | `OQT_GEOM_SIZE_LIMIT`     | `geom_size_limit`       | `100`                                               | Area restriction of the input geometry to the OQT API (sqkm)                 |
+| Python Log Level            | `OQT_LOG_LEVEL`           | `log_level`             | `INFO`                                              | Python logging level                                                         |
+| Semaphore                   | `OQT_SEMAPHORE`           | `semaphore`             | `4`                                                 | Limit number of concurrent Indicator computations for one API request        |
```
Its not really just the limitation on concurrent request to the ohsome API, but a limitation on how many indicators are computed concurrently. Of course the requests to the ohsome API is where concurrency plays the biggest role.

`ohsome_api_limit` does not give hints about how the limit is archived (e.g. semaphore). I prefer just semaphore. This makes it clear that asynchronicity is at play and configurable.

joker234 replied:
Something with `concurrency` would be nice.

joker234 commented on CHANGELOG.md:
```diff
@@ -2,23 +2,33 @@
 
 ### New Features
 
 -  Add support for `groupBy/boundary` queries to the ohsome API client ([#272])
+- Configure OQT using files or environment variables ([#255])
```
```suggestion
- Add support for `groupBy/boundary` queries to the ohsome API client ([#272])
- Configure OQT using files or environment variables ([#255])
```
